### DISCUSSION TITLE
[Feature] content-based visual information retrieval (CBVIR)

### DISCUSSION THREAD

**Discussion by Utopiah** [immich-app/immich#2891] [Feature Request]
### The feature

I discussed yesterday with colleagues about searching through large image dataset. I skimmed today through https://neptune.ai/blog/content-based-image-retrieval-with-siamese-networks and wondering if it could help beyond going through explicit categories like microsoft/resnet-50 or clip-ViT-B-32 provide.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by bo0tzz** [comment]
Immich already uses CLIP for search.

**Reply by Utopiah** [reply]
Doesn't CLIP (which I mentioned as `clip-ViT-B-32`) relies on named concepts and thus wouldn't be as flexible as CBVIR?

**Reply by bo0tzz** [reply]
No, CLIP maps sentences and images into a shared embedding space and is really flexible. You can search for basically anything :)

**Reply by Utopiah** [reply]
Thanks for the clarification and sorry to bother you if you already answered this and I didn't understand but doesn't this mean that named concepts, or sentences, or keywords, must be used for queries whereas CBVIR allow to search for similarity of the entire or only part of the image? Can this still be done via CLIP?

**Reply by bo0tzz** [reply]
CLIP can search based on any query. You can use a single word, a sentence, or an elaborate description and CLIP should be able to find images based on any of those. There's no restriction to certain concepts or keywords.

**Reply by Utopiah** [reply]
Yes but a textual query, one can not e.g select the entire image or part of it, e.g its blue sky, and find similar images, which is what CBVIR would allow.

**Reply by jrasm91** [reply]
Clip can do this already.

**Reply by Utopiah** [reply]
Can you please provide an example of CLIP where the query is not text based but rather part of an image? I didn't even notice that in the interface but maybe I used an old version.

**Reply by alextran1502** [reply]
![image](https://github.com/immich-app/immich/assets/27055614/b3ef537c-e74a-4ae6-a22f-0fd041fc3a43)
 
Is this what you mean?

**Reply by Utopiah** [reply]
That's what I don't mean, namely not typing text to get similar things based on description yet visually clearly very different.

**Reply by jrasm91** [reply]
A query is just a closest neighbor calculation, which you can do with any set of vectors and a vector input. Clip maps both text and photos to a vector, which is why the text query works. I don't see any reason, from a technical perspective, why this is not possible to do with an image as in input.

**Reply by Utopiah** [reply]
Technically yes but AFAIK going image to text to then query will not bring visual similarity.

**Reply by jrasm91** [reply]
We would go image to vector. Not image to text to vector.

**Reply by jrasm91** [reply]
I think you might be hinting at the _accuracy_ differences between the models. Clip is obviously great at certain things, so it would make sense to validate the accuracy of its image similarity capabilities, before considering whether to use an alternative model.

**Reply by Utopiah** [reply]
Point being that, and maybe I'm not using the right description for the implementation, the example with "green tree blue sky" provided above is not the same as taking the first photo and asking for that specific tree without having to type a description for it.

**Reply by Utopiah** [reply]
I'm also NOT suggesting to drop CLIP but rather asking how to extend search.

**Reply by alextran1502** [reply]
Now you got me curious, can you show an example use case?

**Reply by bo0tzz** [reply]
What you're asking for is to take an image and search for similar images, right? CLIP can do this, and it doesn't need to use a text description to do so.

**Comment by Utopiah** [comment]
Related news https://blog.google/products/search/google-circle-to-search-android/
